level_zero_runtime/lib/level_zero_runtime.cpp

```diff
@@ -0,0 +1,296 @@
+                                     gpu_l0_queue->ze_device_, &ret));
+  } else {
+    devDesc.flags = ZE_DEVICE_MEM_ALLOC_FLAG_BIAS_INITIAL_PLACEMENT;
+    CHECK_ZE_RESULT(zeMemAllocDevice(gpu_l0_queue->ze_context_, &devDesc, size,
+                                     alignment, gpu_l0_queue->ze_device_,
+                                     &ret));
+  }
+  return ret;
+}
+
+void dealloc_device_memory(void *queue, void *ptr) {
+  zeMemFree(static_cast<GPU_L0_QUEUE *>(queue)->ze_context_, ptr);
+}
+
+ze_module_handle_t loadModule(void *queue, const void *data, size_t dataSize) {
+  assert(data);
+  auto gpu_l0_queue = static_cast<GPU_L0_QUEUE *>(queue);
+  ze_module_handle_t ze_module;
+  ze_module_desc_t desc = {};
+  desc.format = ZE_MODULE_FORMAT_IL_SPIRV;
+  desc.pInputModule = static_cast<const uint8_t *>(data);
+  desc.inputSize = dataSize;
+  CHECK_ZE_RESULT(zeModuleCreate(gpu_l0_queue->ze_context_,
+                                 gpu_l0_queue->ze_device_, &desc, &ze_module,
+                                 nullptr));
+  return ze_module;
+}
+
+ze_kernel_handle_t getKernel(void *queue, void *module, const char *name) {
+  assert(module);
+  assert(name);
+  ze_kernel_desc_t desc = {};
+  ze_kernel_handle_t ze_kernel;
+  desc.pKernelName = name;
+  CHECK_ZE_RESULT(zeKernelCreate(static_cast<ze_module_handle_t>(module), &desc,
+                                 &ze_kernel));
+  return ze_kernel;
+}
+
+void launchKernel(void *queue, ze_kernel_handle_t kernel, size_t gridX,
+                  size_t gridY, size_t gridZ, size_t blockX, size_t blockY,
+                  size_t blockZ, size_t sharedMemBytes, ParamDesc *params) {
+  assert(kernel);
+  auto paramsCount = countUntil(params, ParamDesc{nullptr, 0});
+
+  auto castSz = [](size_t val) { return static_cast<uint32_t>(val); };
+
+  CHECK_ZE_RESULT(zeKernelSetGroupSize(kernel, castSz(blockX), castSz(blockY),
+                                       castSz(blockZ)));
+  for (size_t i = 0; i < paramsCount; ++i) {
+    auto param = params[i];
+    CHECK_ZE_RESULT(zeKernelSetArgumentValue(kernel, static_cast<uint32_t>(i),
+                                             param.size, param.data));
+  }
+
+  ze_group_count_t launchArgs = {castSz(gridX), castSz(gridY), castSz(gridZ)};
+  auto launch_event = static_cast<GPU_L0_QUEUE *>(queue)->ze_launch_event_;
+  CHECK_ZE_RESULT(zeCommandListAppendLaunchKernel(
+      static_cast<GPU_L0_QUEUE *>(queue)->ze_commandList_, kernel, &launchArgs,
+      launch_event, 0, nullptr));
+  CHECK_ZE_RESULT(zeEventHostSynchronize(launch_event, UINT64_MAX));
+}
+
+// Wrappers
+extern "C" LEVEL_ZERO_RUNTIME_EXPORT void *gpuCreateStream(void *device,
+                                                           void *context) {
+  return catchAll([&]() {
+    if (!device && !context) {
+      return new GPU_L0_QUEUE();
+    } else if (device && context) {
+      return new GPU_L0_QUEUE(static_cast<ze_device_type_t *>(device),
```

Comment:
I think here you want to detect whether you pass a garbage pointer as device or context.  The check might not be complete, but I think there are usually some check like this  at the runtime. Not only depends on compiler to check the type.

Reply:
Can l add something like that later in subsequent PR?. I need to find out if it's possible to do that. I don't see any API's that these libraries provide that I can use.

Reply:
you may reach out to SYCL/L0 runtime to get suggestion.  That's fine you leave it as now but you may put a comments.